lib/chef/data_collector.rb

```diff
@@ -0,0 +1,342 @@
+             Net::HTTPHeaderSyntaxError, Net::ProtocolError, OpenSSL::SSL::SSLError => e
+        disable_data_collector_reporter
+        code = if e.respond_to?(:response) && e.response.code
+                 e.response.code.to_s
+               else
+                 "Exception Code Empty"
+               end
+
+        msg = "Error while reporting run start to Data Collector. " \
+              "URL: #{data_collector_server_url} " \
+              "Exception: #{code} -- #{e.message} "
+
+        if Chef::Config[:data_collector][:raise_on_failure]
+          Chef::Log.error(msg)
+          raise
+        else
+          Chef::Log.warn(msg)
+        end
+      end
+
+      def send_to_data_collector(message)
+        return unless data_collector_accessible?
+
+        Chef::Log.debug("data_collector_reporter: POSTing the following message to #{data_collector_server_url}: #{message}")
+        http.post(nil, message, headers)
+      end
+
+      #
+      # Send any messages to the DataCollector endpoint that are necessary to
+      # indicate the run has completed. Currently, two messages are sent:
+      #
+      # - An "action" message with the node object indicating it's been updated
+      # - An "run_converge" (i.e. RunEnd) message with details about the run,
+      #   what resources were modified/up-to-date/skipped, etc.
+      #
+      # @param opts [Hash] Additional details about the run, such as its success/failure.
+      #
+      def send_run_completion(opts)
+        # If run_status is nil we probably failed before the client triggered
+        # the run_started callback. In this case we'll skip updating because
+        # we have nothing to report.
+        return unless run_status
+
+        send_to_data_collector(Chef::DataCollector::Messages.node_update_message(run_status).to_json)
+        send_to_data_collector(
+          Chef::DataCollector::Messages.run_end_message(
+            run_status: run_status,
+            expanded_run_list: expanded_run_list,
+            total_resource_count: resource_count,
+            updated_resources: updated_resources,
+            status: opts[:status],
+            error_descriptions: error_descriptions
+          ).to_json
+        )
+      end
+
+      def headers(additional_headers = {})
+        headers = { "Content-Type" => "application/json" }
+        headers["x-data-collector-token"] = data_collector_token unless data_collector_token.nil?
+
+        headers.merge(additional_headers)
+      end
+
+      def data_collector_server_url
+        Chef::Config[:data_collector][:server_url]
+      end
+
+      def data_collector_token
+        Chef::Config[:data_collector][:token]
+      end
+
+      def increment_resource_count
+        @resource_count += 1
+      end
+
+      def add_updated_resource(resource_report)
+        @updated_resources << resource_report
+      end
+
+      def disable_data_collector_reporter
+        # TODO: should the event dispatcher support de-registering?
```

Comment:
If we have trouble reaching the DataCollector endpoint at the start of the run, we make the decision to not try and reach it at the end of the run either. However, throughout the run, The Event Dispatcher is still sending events to this instance, which is collecting data and stuff, and then at the end of the run, just decides not to send it.

The thought was perhaps it would be nice if there was a way for a event handler to unregister itself so the Chef run isn't still invoking it for all the events triggered. It's a nice-to-have, but not a necessity. I can remove the TODO for clarity as it's out-of-scope for this change if you think that's prudent.